### PR TITLE
Add cclib --> horton 3 bridge, along with tests

### PR DESCRIPTION
**Pull request by mscho527**
Related Issue : #880

This PR is related to Issue #880 and builds upon #884 . This PR adds a bridging function that constructs Horton 3 IOData object from ccData object.

Test compares the parsed results (iodata.api.load_one) and converted results (horton 3 IOData object converted from ccData object using the bridge function).

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -28,17 +28,36 @@ before_install:
...
 
         source activate p4env
         conda install pytest pytest-cov codecov -c conda-forge
         pip install pytest-shutil
         conda list
+        pip install git+https://github.com/theochem/iodata.git
     else
         python -m pip install -U pytest pytest-cov
     fi
+    if [[ $(echo "$TRAVIS_PYTHON_VERSION < 3.0" | bc -l) == 1 ]];
```
Comment (langner) [inline .travis.yml:41]: This is all the same as in #873 - could you separate it out into a smaller separate PR?

Comment (mscho527) [reply .travis.yml:41]: @langner I am a bit confused here -- I thought that the changes in Travis should be included in the PR so that there can be tests run for the new methods? If you meant that the Travis configurations / test functions should be a separate PR and the bridge functions in another PR, I can definitely work it out tomorrow. Thanks!

Comment (langner) [reply .travis.yml:41]: Yeah, if it's something that's useful in both PRs and can exist in master before either of them, why not merge it in earlier? Trimming down your PRs is always good.

File: cclib/bridge/cclib2horton.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    if (found_iodata):
+        return 3
+    elif (found_horton):
+        return 2
+
+
+def makehorton(ccdat):
+    """ Create horton IOData object from ccData object """
+
+    _check_horton(_found_horton, _found_iodata, _old_horton)
```
Comment (langner) [inline cclib/bridge/cclib2horton.py:46]: Don't these variables live at module level? You don't need to pass them in as arguments.

File: cclib/bridge/cclib2horton.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        if (hasattr(ccdat, 'coreelectrons') and isinstance(ccdat.coreelectrons, numpy.ndarray)):
+            attributes['atcorenums'] = ccdat.coreelectrons
+        # if mulliken charges are known; dict of numpy.ndarrays (size natom)
+        if (hasattr(ccdat, 'atomcharges') and isinstance(ccdat.atomcharges, dict)):
+            attributes['atcharges'] = ccdat.atomcharges
+
+    return IOData(**attributes)
+
+
+def makecclib(iodat):
```
Comment (langner) [inline cclib/bridge/cclib2horton.py:99]: Another thing you could have done is put makehorton and makecclib in separate PRs.

File: cclib/bridge/cclib2horton.py (hunk 2 above)
Comment (mscho527) [reply cclib/bridge/cclib2horton.py:46]: Yes those variables do live at module level. I've wrote the function this way to be consistent with other bridges, but I can modify this to make it more readable and efficient. Thanks!

File: cclib/bridge/cclib2horton.py [hunk 4]
```diff
@@ -82,7 +82,40 @@ def makehorton(ccdat):
             attributes["npa_charges"] = ccdat.atomcharges["natural"]
 
     elif hortonver == 3:
-        pass  # Populate with bridge for horton 3 in later PR
+        # In horton 3 IOData, inputs are type-checked -- thus the bridge also verifies the types.
+        # if coordinates are known; numpy.ndarray (size natom-by-3)
+        if hasattr(ccdat, "atomcoords") and isinstance(ccdat.atomcoords, numpy.ndarray):
```
Comment (berquist) [inline cclib/bridge/cclib2horton.py:87]: Let's use `np.asanyarray` to do conversions throughout again, otherwise these attributes will silently get ignored when you really want them to be added.

Though, at the point you have a `ccData` instance, all of these should already be `ndarray`s. so I wonder if any of that is necessary.

File: cclib/bridge/cclib2horton.py [hunk 5]
```diff
@@ -82,7 +82,40 @@ def makehorton(ccdat):
...
+                kind="restricted",
+                norba=len(ccdat.mocoeffs[0]),
+                coeffs=ccdat.mocoeffs[0].T,
+            )
+            # and if unrestricted:
+            if len(mocoeffs) == 2:
+                attributes["mo"].kind = "unrestricted"
+                attributes["mo"].norbb = len(ccdat.mocoeffs[1])
+                attributes["mo"].coeffs.append(ccdat.mocoeffs[1].T)
+        # if multiplicity known; float / should not be set when mocoeffs present
```
Comment (berquist) [inline cclib/bridge/cclib2horton.py:104]: Can you explain why?

File: cclib/bridge/cclib2horton.py (hunk 2 above)
Comment (langner) [reply cclib/bridge/cclib2horton.py]: Alright, thanks for being consistent.

File: cclib/bridge/cclib2horton.py (hunk 5 above)
Comment (mscho527) [reply cclib/bridge/cclib2horton.py:104]: I think it's meant so that the multiplicity is derived from MO and rather not be overridden by the `spinpol` attribute. This is mentioned in `IOData` [code](https://github.com/theochem/iodata/blob/b36513d162f99b57264005583701c6987037839c/iodata/iodata.py#L174). I'll add this as comment for reference. Thanks.

File: cclib/bridge/cclib2horton.py (hunk 4 above)
Comment (mscho527) [reply cclib/bridge/cclib2horton.py]: I've removed the unnecessary checks you have mentioned. Thanks for the comment!

File: cclib/bridge/cclib2horton.py [hunk 6]
```diff
@@ -137,17 +183,18 @@ def makecclib(iodat):
...
             if iodat.mo.kind == "unrestricted":
                 attributes["mocoeffs"].append(iodat.mo.coeffs[iodat.mo.norba :].T)
-        if hasattr(iodat, "spinpol"):
+        if hasattr(iodat, "spinpol") and isinstance(iodat.spinpol, int):
             # IOData stores 2S, ccData stores 2S+1.
             attributes["mult"] = iodat.spinpol + 1
         if hasattr(iodat, "atnums"):
+            attributes["atnums"] = numpy.asanyarray(iodat.atnums)
+        if hasattr(iodat, "atcorenums") and isinstance(iodat.atnums, numpy.ndarray):
             # cclib stores num of electrons screened out by pseudopotential
```
Comment (mscho527) [inline cclib/bridge/cclib2horton.py:192]: These comments were placed in a wrong (a block above) location in the previous PR. This should fix it.

File: cclib/bridge/cclib2horton.py [hunk 7]
```diff
@@ -80,7 +80,53 @@ def makehorton(data):
...
+                "occs": numpy.concatenate(
+                    numpy.ones(data.homos[0]), numpy.zeros(len(data.mocoeffs[0]) - data.homos[0])
+                ),
+                "coeffs": data.mocoeffs[0],
+                "energies": None,
+                "irreps": None,
+            }
+            # and if unrestricted:
+            if len(mocoeffs) == 2:
+                moattr.kind = "unrestricted"
```
Comment (berquist) [inline cclib/bridge/cclib2horton.py:105]: Shouldn't these be key access rather than attribute access?

File: cclib/bridge/cclib2horton.py [hunk 8]
```diff
@@ -80,7 +80,53 @@ def makehorton(data):
...
+        # Refer to IOData code:
+        # https://github.com/theochem/iodata/blob/b36513d162f99b57264005583701c6987037839c/iodata/iodata.py#L174
+        if (
+            hasattr(data, "mult")
+            and data.mult != None
+            and not isinstance(data.mocoeffs, numpy.ndarray)
+        ):
+            attributes["spinpol"] = data.mult - 1  # horton has 2S+1, iodata has 2S
+        # if pseudopotentials exist; numpy.ndarray (size natom)
+        if hasattr(data, "coreelectrons") and isinstance(data.coreelectrons, numpy.ndarray):
```
Comment (berquist) [inline cclib/bridge/cclib2horton.py:125]: `asanyarray` trick?

File: cclib/bridge/cclib2horton.py [hunk 9]
```diff
@@ -80,7 +80,53 @@ def makehorton(data):
...
+                    )
+                )
+            attributes["mo"] = MolecularOrbitals(**moattr)
+        # if multiplicity known; float / should not be set when mocoeffs present
+        # Refer to IOData code:
+        # https://github.com/theochem/iodata/blob/b36513d162f99b57264005583701c6987037839c/iodata/iodata.py#L174
+        if (
+            hasattr(data, "mult")
+            and data.mult != None
+            and not isinstance(data.mocoeffs, numpy.ndarray)
```
Comment (berquist) [inline cclib/bridge/cclib2horton.py:121]: Should this be a `not hasattr` instead?

File: cclib/bridge/cclib2horton.py [hunk 10]
```diff
@@ -80,7 +80,53 @@ def makehorton(data):
...
+                        numpy.zeros(len(data.mocoeffs[1]) - data.homos[1]),
+                    )
+                )
+            attributes["mo"] = MolecularOrbitals(**moattr)
+        # if multiplicity known; float / should not be set when mocoeffs present
+        # Refer to IOData code:
+        # https://github.com/theochem/iodata/blob/b36513d162f99b57264005583701c6987037839c/iodata/iodata.py#L174
+        if (
+            hasattr(data, "mult")
+            and data.mult != None
```
Comment (berquist) [inline cclib/bridge/cclib2horton.py:120]: I think `mult` will always be an integer when present, unless you've found otherwise.

File: test/bridge/testhorton.py [hunk 11]
```diff
@@ -107,7 +107,34 @@ def test_makehorton(self):
                     )
 
         elif self._hortonver == 3:
-            pass
+            # Identify attributes that should be verified
+            check = ["spinpol"]  # float or int
+            checkArr = ["atcoords", "atnums", "atcorenums"]  # one dimensional arrays
+
+            # IOData in horton 3 uses attr and initializes all possible attributes by default value of None. Thus, hasattr cannot be used here by its own.
```
Comment (berquist) [inline test/bridge/testhorton.py:114]: Can you reflow this comment?

File: cclib/bridge/cclib2horton.py (hunk 10 above)
Comment (langner) [reply cclib/bridge/cclib2horton.py:120]: Yep, you can assume the types specified in [data.py](https://github.com/cclib/cclib/blob/master/cclib/parser/data.py#L103) will hold.

File: cclib/bridge/cclib2horton.py [hunk 12]
```diff
@@ -80,7 +80,53 @@ def makehorton(data):
...
+            and not isinstance(data.mocoeffs, numpy.ndarray)
+        ):
+            attributes["spinpol"] = data.mult - 1  # horton has 2S+1, iodata has 2S
+        # if pseudopotentials exist; numpy.ndarray (size natom)
+        if hasattr(data, "coreelectrons") and isinstance(data.coreelectrons, numpy.ndarray):
+            attributes["atcorenums"] = data.atomnos - data.coreelectrons
+        # if mulliken charges are known; dict of numpy.ndarrays (size natom)
+        if hasattr(data, "atomcharges") and isinstance(data.atomcharges, dict):
```
Comment (langner) [inline cclib/bridge/cclib2horton.py:128]: Another redundant check.

File: test/bridge/testhorton.py [hunk 13]
```diff
@@ -107,7 +107,34 @@ def test_makehorton(self):
                     )
 
         elif self._hortonver == 3:
```
Comment (langner) [inline test/bridge/testhorton.py:109]: Can you please make a separate test class for 2 and 3 versions?

File: cclib/bridge/cclib2horton.py (hunk 7 above)
Comment (mscho527) [reply cclib/bridge/cclib2horton.py:105]: As far as I know, these indeed are attribute access since an instance of horton3's `MolecularOrbital` class needs to be created beforehand. [This ](https://github.com/theochem/iodata/blob/b36513d162f99b57264005583701c6987037839c/iodata/orbitals.py#L30)is the code block that is relevant. I will recheck this soon to make sure.

Comment (langner) [reply cclib/bridge/cclib2horton.py:105]: But in the previous expression `moattr` is created as a dictionary.

Comment (mscho527) [reply cclib/bridge/cclib2horton.py:105]: My bad -- sorry for the confusion. I'll modify this. Thanks for catching this!

File: cclib/bridge/cclib2horton.py (hunk 8 above)
Comment (mscho527) [reply cclib/bridge/cclib2horton.py]: Yes -- I've swapped the check with `asanyarray`. Thanks!

File: cclib/bridge/cclib2horton.py (hunk 9 above)
Comment (mscho527) [reply cclib/bridge/cclib2horton.py]: Thanks for finding this! I've changed it to `not hasattr`.

File: cclib/bridge/cclib2horton.py (hunk 10 above)
Comment (mscho527) [reply cclib/bridge/cclib2horton.py]: I've removed the unnecessary type checks. Thank you!

File: test/bridge/testhorton.py (hunk 11 above)
Comment (mscho527) [reply test/bridge/testhorton.py]: I've reflowed it for line length of 100. Thanks.

File: cclib/bridge/cclib2horton.py (hunk 12 above)
Comment (mscho527) [reply cclib/bridge/cclib2horton.py]: I've removed the unnecessary checks. Thanks.

File: test/bridge/testhorton.py (hunk 13 above)
Comment (mscho527) [reply test/bridge/testhorton.py]: I've separated horton 2 and 3 test classes. Thanks!

**Comment by mscho527** [comment]
Added a comment about `atomcharges` being a dictionary and rebased against master

File: cclib/bridge/cclib2horton.py [hunk 14]
```diff
@@ -80,7 +80,52 @@ def makehorton(data):
...
+        # if atomic numbers are known; numpy.ndarray (size natom)
+        if hasattr(data, "atomnos"):
+            attributes["atnums"] = numpy.asanyarray(data.atomnos)
+        # if orbital coefficients known; iodata.orbitals.MolecularOrbitals
+        if hasattr(data, "mocoeffs"):
+            moattr = {
+                "kind": "restricted",
+                "norba": len(data.mocoeffs[0]),
+                "norbb": None,
+                "occs": numpy.concatenate(
```
Comment (langner) [inline cclib/bridge/cclib2horton.py:96]: Could you add a comment here explaining what this is?

File: cclib/bridge/cclib2horton.py [hunk 15]
```diff
@@ -80,7 +80,52 @@ def makehorton(data):
...
+                    numpy.ones(data.homos[0]), numpy.zeros(len(data.mocoeffs[0]) - data.homos[0])
```
Comment (langner) [inline cclib/bridge/cclib2horton.py:97]: Isn't `len(data.mocoeffs[0]) == data.nbasis`?

File: cclib/bridge/cclib2horton.py (hunk 14 above)
Comment (mscho527) [reply cclib/bridge/cclib2horton.py:101]: I've added a comment about the MolecularOrbital class and the structure of `occs`.

File: cclib/bridge/cclib2horton.py (hunk 15 above)
Comment (mscho527) [reply cclib/bridge/cclib2horton.py]: Yes. Thank for pointing that out. I've just replaced it.
